internal/controlplane/handlers_profile.go

```diff
@@ -644,3 +554,256 @@ func (s *Server) UpdateProfile(ctx context.Context,
+	}
+
+	resp, err := s.processProfileStatus(ctx, dbProfileStatus)
+	if err != nil {
+		return nil, err
+	}
+
+	return &minderv1.GetProfileStatusByIdResponse{
+		ProfileStatus:        resp.ProfileStatus,
+		RuleEvaluationStatus: resp.RuleEvaluationStatus,
+	}, nil
+}
+
+func (s *Server) processProfileStatus(
+	ctx context.Context,
+	dbProfileStatus interface{}) (*minderv1.GetProfileStatusResponse, error) {
+	// Extract Profile Details
+	profileName, profileID, lastUpdated, profileStatus, extractProfileDetailsErr := extractProfileDetails(dbProfileStatus)
+
+	if extractProfileDetailsErr != nil {
+		return nil, extractProfileDetailsErr
+	}
+
+	// Prepare rule evaluation filter
+	var ruleEvaluationStatuses []*minderv1.RuleEvaluationStatus
+
+	// Extract filter criteria based on request type
+	selector, ruleType, ruleName, extractFiltersErr := extractFiltersFromRequest(ctx.Value(requestKey))
+
+	if extractFiltersErr != nil {
+		return nil, extractFiltersErr
+	}
+
+	// Retrieve rule evaluation statuses if needed
+	if selector != nil || isAllRequested(ctx) {
+		dbRuleEvaluationStatuses, err := s.store.ListRuleEvaluationsByProfileId(ctx, db.ListRuleEvaluationsByProfileIdParams{
+			ProfileID:    profileID,
+			EntityID:     *selector,
+			RuleTypeName: *ruleType,
+			RuleName:     *ruleName,
+		})
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return nil, status.Errorf(codes.Unknown, "failed to list rule evaluation status: %s", err)
+		}
+
+		ruleEvaluationStatuses = s.getRuleEvaluationStatuses(
+			ctx, dbRuleEvaluationStatuses, profileID.String(),
+		)
+	}
+
+	// Telemetry logging
+	entityCtx := engcontext.EntityFromContext(ctx)
+	logger.BusinessRecord(ctx).Project = entityCtx.Project.ID
+	logger.BusinessRecord(ctx).Profile = logger.Profile{Name: profileName, ID: profileID}
+
+	return &minderv1.GetProfileStatusResponse{
+		ProfileStatus: &minderv1.ProfileStatus{
+			ProfileId:     profileID.String(),
+			ProfileName:   profileName,
+			ProfileStatus: profileStatus,
+			LastUpdated:   lastUpdated,
+		},
+		RuleEvaluationStatus: ruleEvaluationStatuses,
+	}, nil
+}
+
+// Helper function to extract entity selector
+func extractEntitySelector(entity *minderv1.EntityTypedId) *uuid.NullUUID {
+	if entity == nil {
+		return nil
+	}
+	var selector uuid.NullUUID
+	if err := selector.Scan(entity.GetId()); err != nil {
+		return nil
+	}
+	return &selector
+}
+
+// Helper Function Check if all items are requested
+func isAllRequested(ctx context.Context) bool {
+	switch req := ctx.Value("request").(type) {
+	case *minderv1.GetProfileStatusByNameRequest:
+		return req.GetAll()
+	case *minderv1.GetProfileStatusByIdRequest:
+		return req.GetAll()
+	}
+	return false
+}
+
+func extractProfileDetails(dbProfileStatus interface{}) (
+	profileName string,
+	profileID uuid.UUID,
+	lastUpdated *timestamppb.Timestamp,
+	profileStatus string,
+	err error) {
+
+	switch ps := dbProfileStatus.(type) {
+	case db.GetProfileStatusByNameAndProjectRow:
```

Comment:
The isn't much benefit in having a common helper function if it has switch statements which separate the code paths. 
In this case, we can change the signature of `processProfileStatus` to `processProfileStatus(ctx context.Context, profileName string, profileID uuid.UUID, lastUpdated *timestamppb.Timestamp, profileStatus string)` and then just pass the fields that we need:
```
	resp, err := s.processProfileStatus(
		ctx,
		dbProfileStatus.Name,
		dbProfileStatus.ID,
		timestamppb.New(dbProfileStatus.LastUpdated),
		string(dbProfileStatus.ProfileStatus),
	)
```